tests_integration/nativeapp/test_command_metric_counters.py

```diff
@@ -0,0 +1,248 @@
+# Copyright (c) 2024 Snowflake Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from shlex import split
+from typing import Dict, Callable
+from unittest import mock
+
+from snowflake.cli._app.telemetry import TelemetryEvent, CLITelemetryField
+from snowflake.cli.api.metrics import CLICounterField
+from tests.nativeapp.factories import (
+    ProjectV11Factory,
+    ProjectV2Factory,
+    ApplicationPackageEntityModelFactory,
+    ApplicationEntityModelFactory,
+)
+from tests.project.fixtures import *
+from tests_common import temp_dir
+from tests_integration.test_utils import extract_first_telemetry_message_of_type
+
+
+@pytest.mark.integration
+@pytest.mark.parametrize(
+    "command,expected_counter",
+    [
+        (
+            [
+                "sql",
+                "-q",
+                "select '<% ctx.env.test %>'",
+                "--env",
+                "test=value_from_cli",
+            ],
+            1,
+        ),
+        (["sql", "-q", "select 'string'"], 0),
+    ],
+)
+@mock.patch("snowflake.connector.telemetry.TelemetryClient.try_add_log_to_batch")
+def test_sql_templating_emits_counter(
+    mock_telemetry,
+    command: List[str],
+    expected_counter,
+    runner,
+):
+    result = runner.invoke_with_connection_json(command)
+
+    assert result.exit_code == 0
+
+    message = extract_first_telemetry_message_of_type(
+        mock_telemetry, TelemetryEvent.CMD_EXECUTION_RESULT.value
+    )
+
+    assert (
+        message[CLITelemetryField.COUNTERS.value][CLICounterField.SQL_TEMPLATES]
+        == expected_counter
+    )
+
+
+@pytest.mark.integration
+@pytest.mark.parametrize(
+    "command,project_factory,expected_counters",
+    [
+        (
+            "app deploy",
+            lambda: ProjectV11Factory(
```

Comment:
This is the best I could come up with to have parametrization with factories, if anyone has a better approach I'd be interested

Reply:
what do we need parametrized tests at all? I don't think I can read this anymore. I think overall we make way, way too frequent use of parametrized tests in our codebase, but that's a different story.

Reply:
if the arguments to the factories are the same, you can parametrize just on the class and instantiate it in the test
```python
@pytest.mark.parametrize("factory_class", [FooFactory, BarFactory])
def test_factory(factory_class):
    obj = factory_class(arg=123)
```

Reply:
I ended up getting rid of all the parametrization and separating the tests out since the factories didn't have the same arguments, I think it's more clear and separated this way, let me know what you think